### ISSUE TITLE
Error messages are never sent to the DLQ

### ISSUE THREAD

**Issue by antoin-m** [spring-cloud/spring-cloud-stream-binder-kafka#657]
Hi,

I created a Spring Cloud Stream application using the Kafka Binder, here is my yml configuration file:
```
spring:
  cloud:
    stream:
      kafka:
        binder:
          brokers: localhost
          consumerProperties:
            schema.registry.url: http://localhost:8081
            key.deserializer: org.apache.kafka.common.serialization.StringDeserializer
            value.deserializer: io.confluent.kafka.streams.serdes.avro.SpecificAvroDeserializer
          producerProperties:
            schema.registry.url: http://localhost:8081
            key.serializer: org.apache.kafka.common.serialization.StringSerializer
            value.serializer: io.confluent.kafka.streams.serdes.avro.SpecificAvroSerializer
        bindings:
          input:
            consumer:
              startOffset: earliest
              enableDlq: true
              dlqName: dlq
      bindings:
        input:
          destination: foo
          group: fooGroup
          consumer:
            maxAttempts: 1
            useNativeDecoding: true
        output:
          destination: bar
          producer:
            useNativeEncoding: true
      function:
        definition: processor
```

Everything works great but the DLQ. While inspecting the broker I can see that it gets created yet no messages are sent to the topic.
Here are the different processors I tried:

```
    @Bean
    public Function<EventA, EventB> processor() {
        return input -> {
            throw new RuntimeException("FAIL");
        };
    }
```

```
    @Bean
    public Function<Flux<EventA>, Flux<EventB>> processor() {
        return input -> input.map(i -> {
            throw new RuntimeException("FAIL");
        });
    } 
```

Libraries used:

```
    compile group: 'org.springframework.cloud', name: 'spring-cloud-stream', version: '2.2.0.RELEASE'
    compile group: 'org.springframework.cloud', name: 'spring-cloud-stream-binder-kafka', version: '2.2.0.RELEASE'
    compile group: 'org.springframework.cloud', name: 'spring-cloud-stream-schema', version: '2.2.0.RELEASE'
```

Did I miss something?

**Comment by ddulaney** [comment]
Hi @sobychacko! I ran into this issue as well. Is it possible that the DLQ key/val serializers could be pulled from `producerProperties` instead? It seems strange that we need to specify them in both places.

**Comment by sobychacko** [comment]
@ddulaney Could you share your configuration where it fails? If you share a small reproducible app, then it will help to debug to see if there are any issues. I am reopening this issue.

**Comment by garyrussell** [comment]
>Is it possible that the DLQ key/val serializers could be pulled from producerProperties instead? It seems strange that we need to specify them in both places.

How would the framework decide which producer to select, and for which consumer's DLQ publisher?

Apps can have multiple producers and consumers.

**Comment by ddulaney** [comment]
Thanks @sobychacko! I'll spend some time to pare down my app, but here's my `application.yml`:

```yaml
spring:
  cloud:
    stream:
      # Default configs for all bindings (Kafka agnostic)
      default:
        # Set the content type to Avro
        contentType: application/*+avro
        # Set the Kafka group-id
        group: my-service
        consumer:
          # Inbound messages will be deserialized directly by the Confluent KafkaAvroDeserializer
          # See the key/value.deserializer class values below
          useNativeDecoding: true
        producer:
          # Outbound messages will be serialized directly by the Confluent KafkaAvroSerializer
          # See the key/value.serializer class values below
          useNativeEncoding: true
      # Binding-specific configs (Kafka agnostic)
      bindings:
        input:
          # Topic to consume from
          destination: my-topic
      # Kafka-specific configs
      kafka:
        bindings:
          input:
            consumer:
              # Enable dead letter queue
              enableDlq: true
              # DLQ topic to produce to on error
              dlqName: my-topic.dlq
              # Since we're using native decoding, we need to specify the DLQ serializers here
              dlqProducerProperties.configuration:
                key.serializer: io.confluent.kafka.serializers.KafkaAvroSerializer
                value.serializer: io.confluent.kafka.serializers.KafkaAvroSerializer
        binder:
          # Kafka brokers
          brokers: ${BOOTSTRAP_SERVERS:localhost:9092}
          # Don't auto-create topics
          autoCreateTopics: ${AUTO_CREATE_TOPICS:false}
          # Consumer properties
          consumerProperties:
            schema.registry.url: ${SCHEMA_REGISTRY_URL:http://localhost:8081}
            key.deserializer: io.confluent.kafka.serializers.KafkaAvroDeserializer
            value.deserializer: io.confluent.kafka.serializers.KafkaAvroDeserializer
            # Automatically cast consumed message to specific class rather than a generic message
            specific.avro.reader: true
          # Producer properties (for publishing to DLQ)
          producerProperties:
            schema.registry.url: ${SCHEMA_REGISTRY_URL:http://localhost:8081}
            key.serializer: io.confluent.kafka.serializers.KafkaAvroSerializer
            value.serializer: io.confluent.kafka.serializers.KafkaAvroSerializer
```

@garyrussell - Correct me if I'm wrong, but `consumerProperties` & `producerProperties` here are meant to apply to all consumers & producers for the Kafka binder (unless overridden for a specific binding). My consumer is picking up the schema registry URL, deserializers, etc from `consumerProperties`, and the DLQ producer is picking up the schema registry but **not** the serializers.

**Comment by garyrussell** [comment]
Ah; sorry; I was referring to specific consumer/producer bindings, not the producer properties at the binder level.

Yes, I think this is a reasonable request, but it could be a breaking change so it will probably have to be done in the next release.

**Comment by sobychacko** [comment]
@ddulaney Regarding your suggestion about providing `schema.registry.url` only once through binder configuration may not work without the binder makes certain exceptions. This is because the binder only merges the known consumer configurations into the final property list that goes into the listener container. Since `schema.registry.url` is not such a well known consumer config it gets excluded.

**Comment by ddulaney** [comment]
@sobychacko Gotcha, I thought that might be the case. Thanks!

**Comment by sobychacko** [comment]
@ddulaney The changes for binder producer properties reused for `dlq` are now merged to master (`3.1.0.BUIILD-SNAPSHOT`).

**Comment by ddulaney** [comment]
Wow you guys are fast. Thank you!
